### PR TITLE
Add bucket-related operations 

### PR DESCRIPTION
**Pull request by lynnliu030**
Add bucket-related operations for now: create, delete, list; for the server side, enable registering physical locations to logical buckets in SkyStore

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,12 +41,27 @@ Run some sample commands using the AWS CLI:
...
 just run-cli-multipart
 ```
 
+Test the server
+```
+cd store-server
+just test
+```
+
+Before E2E test, if make changes to the server, then run
```
Comment (simon-mo) [inline README.md:58]:
```suggestion
Before E2E test, if make changes to the server's API, then run the following to re-generate the rust client code. 
```

File: config/init_config.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (simon-mo) [inline config/init_config.json:1]: move theses to a test or example folder

File: s3-proxy/src/.DS_Store
Comment (simon-mo) [inline s3-proxy/src/.DS_Store:1]: remove this. i also added ignoring .DS_Store to the main branch.

File: s3-proxy/src/client_impls/gcp.rs [hunk 3]
```diff
@@ -27,6 +30,68 @@ impl GCPObjectStoreClient {
...
+
+        self.client
+            .get_bucket(&GetBucketRequest {
+                bucket,
+                ..Default::default()
+            })
+            .await
+            .unwrap();
+
+        Ok(S3Response::new(HeadBucketOutput::default()))
```
Comment (simon-mo) [inline s3-proxy/src/client_impls/gcp.rs:92]: This need to be fixed and tested

File: s3-proxy/src/client_impls/azure.rs [hunk 4]
```diff
@@ -132,10 +132,72 @@ impl AzureObjectStoreClient {
...
+        req: S3Request<HeadBucketInput>,
+    ) -> S3Result<S3Response<HeadBucketOutput>> {
+        let req = req.input;
+        let container_name = req.bucket;
+
+        let container_client = self.container_client(&container_name);
+        let resp = container_client.get_properties().await;
+
+        match resp {
+            Ok(_) => Ok(S3Response::new(HeadBucketOutput::default())),
```
Comment (simon-mo) [inline s3-proxy/src/client_impls/azure.rs:194]: This need to be fixed and tested.

File: s3-proxy/src/skyproxy.rs [hunk 5]
```diff
@@ -120,6 +200,142 @@ impl SkyProxy {
             }
         }
     }
+
+    async fn warmup(
```
Comment (simon-mo) [inline s3-proxy/src/skyproxy.rs:204]: wait what does this do?

**Review by simon-mo — CHANGES_REQUESTED** [review]
It seems like this PR are doing many things:
- Bucket operations (not fully done)
- CLI (this is nice but needs a separate design, we can also write it in Python if that's easier?)
- Configuration system so it is not hard coded. 

Let's do smaller PR first?

File: s3-proxy/src/skyproxy.rs (hunk 5 above)
Comment (lynnliu030) [reply s3-proxy/src/skyproxy.rs]: just removed that! used for testing

File: s3-proxy/src/skyproxy.rs [hunk 6]
```diff
@@ -21,23 +22,100 @@ pub struct SkyProxy {
...
+                Err(e) => {
+                    if e.to_string().contains("BucketAlreadyExists") {
+                    } else {
+                        panic!("Failed to create bucket: {}", e);
+                    }
+                }
+            };
         }
 
+        // Real object store configuration (regions: init regions)
```
Comment (simon-mo) [inline s3-proxy/src/skyproxy.rs:69]: Assuming this will be brought back?

File: s3-proxy/src/skyproxy.rs [hunk 7]
```diff
@@ -176,6 +253,174 @@ impl S3 for SkyProxy {
...
+
+        for locator in locators {
+            let client: Arc<Box<dyn ObjectStoreClient>> =
+                self.store_clients.get(&locator.tag).unwrap().clone();
+
+            let bucket_name = locator.bucket.clone();
+            let dir_conf = self.dir_conf.clone();
+
+            // TODO: fix
+            // if buckt name starts with skystore-, then it's a skybucket already created during initialization
```
Comment (simon-mo) [inline s3-proxy/src/skyproxy.rs:285]: Can you explain? When do we create such bucket?

File: s3-proxy/src/skyproxy.rs [hunk 8]
```diff
@@ -176,6 +253,174 @@ impl S3 for SkyProxy {
...
+                )
+                .await
+                .unwrap();
+            });
+        }
+
+        while let Some(result) = tasks.join_next().await {
+            if let Err(err) = result {
+                return Err(s3s::S3Error::with_message(
+                    s3s::S3ErrorCode::InternalError,
```
Comment (simon-mo) [inline s3-proxy/src/skyproxy.rs:328]: Leave a todo to handle cleanup the others if only one failed?

**Review by simon-mo — APPROVED** [review]

File: s3-proxy/src/skyproxy.rs (hunk 6 above)
Comment (lynnliu030) [reply s3-proxy/src/skyproxy.rs:69]: yes, this part is for real object-store testing

File: s3-proxy/src/skyproxy.rs (hunk 7 above)
Comment (lynnliu030) [reply s3-proxy/src/skyproxy.rs:285]: This is creating empty buckets in all init regions to store cache copies later; I guess we can also do bucket creation later when cache copies are created, but this is just for convenience

**Pull request merged**